internal/storage/driver/memory/operation.go

```diff
@@ -190,7 +190,47 @@ func (s *operations) ListOperationsByInstanceID(instanceID string) ([]internal.O
 	return operations, nil
 }
 
-func (s *operations) ListOperationsInTimeRange(time.Time, time.Time) ([]internal.Operation, error) {
+func (s *operations) ListOperationsByInstanceIDGroupByType(instanceID string) (*internal.GroupedOperations, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	grouped := internal.GroupedOperations{
+		ProvisionOperations:      make([]internal.ProvisioningOperation, 0),
+		DeprovisionOperations:    make([]internal.DeprovisioningOperation, 0),
+		UpgradeKymaOperations:    make([]internal.UpgradeKymaOperation, 0),
+		UpgradeClusterOperations: make([]internal.UpgradeClusterOperation, 0),
+		UpdateOperations:         make([]internal.UpdatingOperation, 0),
+	}
+
+	for _, op := range s.operations {
+		switch op.Type {
+		case internal.OperationTypeProvision:
+			grouped.ProvisionOperations = append(grouped.ProvisionOperations, internal.ProvisioningOperation{Operation: op})
+
+		case internal.OperationTypeDeprovision:
+			grouped.DeprovisionOperations = append(grouped.DeprovisionOperations, internal.DeprovisioningOperation{Operation: op})
+
+		case internal.OperationTypeUpgradeKyma:
+			grouped.UpgradeKymaOperations = append(grouped.UpgradeKymaOperations, internal.UpgradeKymaOperation{Operation: op})
+
+		case internal.OperationTypeUpgradeCluster:
+			grouped.UpgradeClusterOperations = append(grouped.UpgradeClusterOperations, internal.UpgradeClusterOperation{Operation: op})
+
+		case internal.OperationTypeUpdate:
+			grouped.UpdateOperations = append(grouped.UpdateOperations, internal.UpdatingOperation{Operation: op})
+		}
+	}
+
+	s.sortProvisioningByCreatedAtDesc(grouped.ProvisionOperations)
```

Comment:
If the input is sorted by CreatedAt the sort here would not be necessary. Operations are already sorted.

Reply:
Because it is in memory implementation the sorting order is not guaranteed and heavily dependent on insertion order in specific test.